charts/k8s-monitoring/templates/agent_config/_processors.river.txt

```diff
@@ -61,9 +61,18 @@ otelcol.processor.k8sattributes "default" {
 otelcol.connector.host_info "default" {
   host_identifiers = [ "k8s.node.name" ]
   output {
-    metrics = [otelcol.processor.batch.batch_processor.input]
+    metrics = [otelcol.processor.batch.host_info_batch.input]
   }
 }
+
+otelcol.processor.batch "host_info_batch" {
+  metrics = [otelcol.exporter.prometheus.host_info_metrics.input]
+}
+
+otelcol.exporter.prometheus "host_info_metrics" {
+  add_metric_suffixes = false
+  forward_to          = [prometheus.remote_write.metrics_service.receiver]
```

Comment:
```suggestion
  forward_to          = [prometheus.relabel.metrics_service.receiver]
```
Using this target instead of directly to the remote_write means that the `cluster` label will get set, and will properly use the desired protocol. (We don't always use remote_write for metrics)

Reply:
We kinda don't want the cluster label set tbh. We will use this metric to bill users, and we bill them based on `count(traces_host_info)`, if there are more labels, there is a likelyhood that the same host is counted twice leading to overcounting.

While `cluster` label might not be causing any issue here we still want to minimise the number of label manipulations to make sure we send the simplest possible metric to Grafana Cloud. 

> We don't always use remote_write for metrics

Can you elaborate on this?

cc @rlankfo

Reply:
Without the cluster label, it becomes more challenging to find out where the source of that metric is. Also, wouldn't you have problems differentiating between sources if you have, say, a dev and a prod cluster with identical deployments (and instrumented applications). The cluster label is what we use to know which cluster the metrics came from.

The `externalServices.prometheus` section allows you to change from `remote_write` to `otlp` or `otlphttp`:
https://github.com/grafana/k8s-monitoring-helm/blob/main/charts/k8s-monitoring/values.yaml#L24

This was a request from a customer to be able to ship metrics via OTLP.

Reply:
If the protocol is otlp, then the `prometheus.remote_write.metrics_service` component wont exist.
The `prometheus.relabel.metrics_service` component is the "fan-in" for all metric sources before being delivered for storage.

Reply:
Okay I see what you mean:

```
Error: /etc/agent/config.river:109:26: component "prometheus.remote_write.metrics_service.receiver" does not exist or is out of scope

108 |   add_metric_suffixes = false
109 |   forward_to          = [prometheus.remote_write.metrics_service.receiver]
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
110 | }

Error: could not perform the initial load successfully
interrupt received
```

I got around it by copying some snippets. Please see my latest commit.

Reply:
The error is from `otelcol.prometheus.receiver` and the following pipeline fails:

```
host_info_connector --> prometheus_exporter --> prometheus_receiver --> otlp output
```

Can we somehow have the following pipeline @petewall :

```
host_info_connector --> otlp output

OR

host_info_connector --> prometheus_exporter --> prometheus_remote_write
```

Reply:
We'll have to do one of the two options presented by Goutham. In this case, we don't want to include any additional labels and the host info connector component should live as close to the remote write component as possible.

Reply:
> Also, wouldn't you have problems differentiating between sources if you have, say, a dev and a prod cluster with identical deployments (and instrumented applications). The cluster label is what we use to know which cluster the metrics came from.

In this case, I don't think care to differentiate by cluster. If customers are sending data to a production cloud stack, we will charge them per host, regardless if that host is part of a dev cluster or not.

Reply:
> > Also, wouldn't you have problems differentiating between sources if you have, say, a dev and a prod cluster with identical deployments (and instrumented applications). The cluster label is what we use to know which cluster the metrics came from.
> 
> In this case, I don't think we care to differentiate by cluster. If customers are sending data to a production cloud stack, we will charge them per host, regardless if that host is part of a dev cluster or not.

I guess my point is not that you necessarily want to differentiate between clusters, the point is to ensure that two different sources of metrics don't conflict. Node names are not guaranteed to be globally unique.

If you have two clusters, dev and prod, and each have nodes name "cluster-node-1", and "cluster-node-2", how do you know the difference between a metric tagged with "cluster-node-1" on dev and "cluster-node-1" on prod?

Reply:
> Can we somehow have the following pipeline

Sure, just move the `{{- if ... }}` conditional that checks for the output protocol into the `forward_to` section of the `otelcol.connector.host_info` or `otelcol.processor.batch` components.

Reply:
> If you have two clusters, dev and prod, and each have nodes name "cluster-node-1", and "cluster-node-2", how do you know the difference between a metric tagged with "cluster-node-1" on dev and "cluster-node-1" on prod?

this could certainly happen but I think it would be an edge case. Typically, the nodes will have different names when using AWS, GCP, Azure, etc.  In this case, we'd likely need the customer to fall back to `host.id` or some other host identifer resource attribute.

Reply:
although thinking about it, adding the cluster label wouldn't increase the cardinality so we'd still get accurate host counts with our custom tracker so we can probably add it but we need to avoid adding any labels that will increase the cardinality.

Reply:
The relabel processor can add more labels based on the config it's given tbh. It's not just cluster label, but I got around the issue for now and everything should be working now.

Could you take a look @petewall